lib/active_record_host_pool/pool_proxy.rb

```diff
@@ -129,7 +129,16 @@ def _connection_pools
     end
 
     def _pool_key
-      @_pool_key ||= "#{@config[:host]}/#{@config[:port]}/#{@config[:socket]}/#{@config[:username]}/#{@config[:slave] && 'slave'}"
+      @_pool_key ||= begin
+        [
+          @config[:host],
+          @config[:port],
+          @config[:socket],
+          @config[:username],
+          @config[:database],
```

Comment:
This change seems intuitive at first, of course you want the name in the cache key, that'll ensure you don't mix up your results right? That's what I thought too but you'll find this change increases the connection count exponentially. You'll have a separate connection for _every_ database, which I think is what this gem was trying to _not_ do. 

> This gem allows for one ActiveRecord connection to be used to connect to multiple databases on a server. It accomplishes this by calling select_db() as necessary to switch databases between database calls.

I believe the solution in #61 is the right solution. It's solves the issue you've described, provides the optimization of Rails 6 and will fail the build when the Rails implementation unexpectedly changes.

E.g. 

```
{"test"=>
  {
   "encoding"=>"utf8",
   "username"=>"puppies",
   "password"=>"puppies-are-cute",
   "adapter"=>"mysql2",
   "host"=>"192.168.1.10",
   "port"=>3306,
   "database"=>"arhp_test_1",
   "shards"=>
    {1001=>{"adapter"=>"mysql2", "host"=>"192.168.1.55", "port"=>3306, "database"=>"test_host_1_db_1_shard_1", "encoding"=>"utf8", "pool"=>5, "hosts"=>["192.168.1.55:3306"]},
     1002=>{"adapter"=>"mysql2", "host"=>"192.168.1.55", "port"=>3306, "database"=>"test_host_1_db_1_shard_2", "encoding"=>"utf8", "pool"=>5, "hosts"=>["192.168.1.55:3306"]},
     1003=>{"adapter"=>"mysql2", "host"=>"192.168.1.55", "port"=>3306, "database"=>"test_host_1_db_1_shard_3", "encoding"=>"utf8", "pool"=>5, "hosts"=>["192.168.1.55:3306"]},
     1004=>{"adapter"=>"mysql2", "host"=>"192.168.1.55", "port"=>3306, "database"=>"test_host_1_db_1_shard_4", "encoding"=>"utf8", "pool"=>5, "hosts"=>["192.168.1.55:3306"]},
     1005=>{"adapter"=>"mysql2", "host"=>"192.168.1.55", "port"=>3306, "database"=>"test_host_1_db_1_shard_5", "encoding"=>"utf8", "pool"=>5, "hosts"=>["192.168.1.55:3306"]},
     1006=>{"adapter"=>"mysql2", "host"=>"192.168.1.55", "port"=>3306, "database"=>"test_host_1_db_1_shard_6", "encoding"=>"utf8", "pool"=>5, "hosts"=>["192.168.1.55:3306"]},
     1007=>{"adapter"=>"mysql2", "host"=>"192.168.1.55", "port"=>3306, "database"=>"test_host_1_db_1_shard_7", "encoding"=>"utf8", "pool"=>5, "hosts"=>["192.168.1.55:3306"]},
     1008=>{"adapter"=>"mysql2", "host"=>"192.168.1.55", "port"=>3306, "database"=>"test_host_1_db_1_shard_8", "encoding"=>"utf8", "pool"=>5, "hosts"=>["192.168.1.55:3306"]},
     1009=>{"adapter"=>"mysql2", "host"=>"192.168.1.55", "port"=>3306, "database"=>"test_host_1_db_1_shard_9", "encoding"=>"utf8", "pool"=>5, "hosts"=>["192.168.1.55:3306"]},
     1010=>{"adapter"=>"mysql2", "host"=>"192.168.1.55", "port"=>3306, "database"=>"test_host_1_db_1_shard_10", "encoding"=>"utf8", "pool"=>5, "hosts"=>["192.168.1.55:31808"]},
    }
  }
}
```

With a configuration like the one above, before this change we would have two database connections, after this change, we'd have one for every shard.

Reply:
Thanks for reviewing @mriddle. 

> You'll have a separate connection for every database, which I think is what this gem was trying to not do.

Yeah, that does seem to this gem's raison d'etre. 

> I believe the solution in #61 is the right solution. It's solves the issue you've described, provides the optimization of Rails 6 and will fail the build when the Rails implementation unexpectedly changes.

With the additional information you shared off Github with how this change could impact real systems in the wild I find myself in agreement. Thanks for reaching out off PR.

One thing worth noting (for posterity and to make sure we are on the same page) is that PR #61 only fixes the known problematic code path and adds tests for that. The issue can still manifest in new ways should Rails or other DB-related gems reference the connection while inside of an operation on a connection adapter instance (e.g. Mysql2Adapter). 

Given that over the lifetime of this gem this has only just now happened with Rails 6 via one code path this seems to keep minimize risk. Should this happen though it will likely be realized in a downstream application where tracking down the cause of the issue may be tricky. 

To keep both the optimization and to eliminate the bug I think it possible, but would require significantly reworking this gem. 

Does PR #61 still feel like the right-sized solution and set of trade-offs?

Reply:
I believe so and instead spend any further significant efforts investing in pushing these changes upstream into Rails since that's the direction it's heading in.